### PR TITLE
Add initial support to the list dtype

### PR DESCRIPTION
**Pull request by philss**
This is the initial work to support "list" datatypes from Arrow/Polars in Explorer.

It may close https://github.com/elixir-explorer/explorer/issues/296.

## Tasks

- [x] `{:list, :integer}`
- [x] `{:list, :boolean}`
- [x] `{:list, :string}`
- [x] `{:list, :float}` (missing special values, like `:nan`) 
- [x] `{:list, :binary}`
- [x] `{:list, :date}`
- [x] `{:list, :time}`
- [x] `{:list, :category}`
- [x] `{:list, {:datetime, _precision}}`
- [x] `{:list, {:duration, _precision}}`
- [x] support creation of homogeneous recursive series - that is: multiple level series, with the same list dtype
- [x] support creation of heterogeneous recursive series - that is: multiple level series, with **different** list dtype
- [x] support inspection of list series
- [x] support casting of list series
- [x] support reads/writes from and to Parquet / IPC / NDJSON files (CSV is not supported). 

---
PS: this is the second attempt to add the list dtype. The first one was: https://github.com/elixir-explorer/explorer/pull/401

### CODE REVIEW

**Comment by philss** [comment]
@josevalim @cigrainger @billylanchantin I think this is ready for review! :D

José asked to be a reviewer, so I requested him, but you all are welcome to review :)

And just to give more context, I had to refactor the "translation" of dtypes because I needed to "parse" the `{:list, _}` dtype that can be recursive. So I moved this translation to the Rust side, and I'm using Rustler features to encode/decode it properly.
Also, I think there are some edge cases in the `check_dtypes!` function - like lists of different depths -, but I chose to not worry about that for now.

**Comment by cigrainger** [comment]
_This is so exciting_. I'll give this a proper read through today.

File: lib/explorer/backend/data_frame.ex [hunk 1]
```diff
@@ -231,56 +231,63 @@ defmodule Explorer.Backend.DataFrame do
...
     ])
   end
 
+  defp to_doc(item, opts) when is_list(item) do
+    open = A.color("[", :list, opts)
+    close = A.color("]", :list, opts)
+    A.container_doc(open, item, close, opts, &to_doc/2)
+  end
+
+  defp to_doc(item, _opts), do: Explorer.Shared.to_string(item)
```
Comment (josevalim) [inline lib/explorer/backend/data_frame.ex:283]: Oh, it seems we only call `Shared.to_string` on inspect. So my suggestion is to move `to_doc` to `Explorer.Shared.to_doc` and merge the to_string implementation into it.

**Review by josevalim — APPROVED** [review]
Beautiful work. Awesome refactoring on the dtype handling. I added just one tiny suggestion and we can ship it.

File: lib/explorer/polars_backend/series.ex [hunk 2]
```diff
@@ -68,31 +51,14 @@ defmodule Explorer.PolarsBackend.Series do
...
-      "datetime[ms]" -> {:s, 64}
-      "datetime[μs]" -> {:s, 64}
-      "datetime[ns]" -> {:s, 64}
-      "duration[ms]" -> {:s, 64}
-      "duration[μs]" -> {:s, 64}
-      "duration[ns]" -> {:s, 64}
-      "cat" -> {:u, 32}
-      dtype -> raise "cannot convert dtype #{inspect(dtype)} to iotype"
-    end
+    Shared.apply_series(series, :s_iotype)
```
Comment (josevalim) [inline lib/explorer/polars_backend/series.ex:61]: No changes required now but, for completeness, we want to support u8, u32, etc as dtypes themselves, which means we will be able to move this function all the way up to `Explorer.Series` and remove the callback altogether in the future. :)

**Review by cigrainger — APPROVED** [review]
Loving the refactor. Tidies things up really nicely! Nothing jumping out at me. :shipit:

File: lib/explorer/backend/data_frame.ex (hunk 1 above)
Comment (philss) [reply lib/explorer/backend/data_frame.ex]: Done! 504531c5fd1b75074d4d1560738dbe3ce231c414

File: lib/explorer/shared.ex [hunk 3]
```diff
@@ -215,32 +224,85 @@ defmodule Explorer.Shared do
...
+
+  @doc """
+  Check if the dtype has the leaf in one of the possible values.
+
+  This is useful to verify if the list dtype of unknown "height" is valid,
+  or if it has the dtype in one of the possibilities.
+  """
+  def leaf_dtype_of?({:list, inner_dtype}, possibilities) do
+    leaf_dtype_of?(inner_dtype, possibilities)
+  end
```
Comment (billylanchantin) [inline lib/explorer/shared.ex:291]: Suggestion: drop this function and make `leaf_dtype` public. This function duplicates the logic of `leaf_dtype`, it's roughly the same number of characters to do `if leaf_dtype(dtype) in list` as it is `if leaf_dtype_of?(dtype, list)`.

**Review by josevalim — APPROVED** [review]

**Review by billylanchantin — APPROVED** [review]
This is _incredible_! In addition to the functionality, which unlocks a few things like `mode` and `explode`, I'm really liking the dtype refactor. It did feel like the string representations of the dtypes were repeated too much.

I left a few suggestions, but they're surface level. Please take 'em or leave 'em!

Also, I started to play with a property test in `test/explorer/series/list_test.exs`, but I don't think it's necessary for this pass. It may be worth coming back to though, especially if we find issues down the road.

File: test/explorer/series/list_test.exs [hunk 4]
```diff
@@ -0,0 +1,286 @@
...
+      s = Series.from_list([[~N[2023-11-10 00:19:30]]])
+
+      assert inspect(s) ==
+               """
+               #Explorer.Series<
+                 Polars[1]
+                 list[datetime[μs]] [[2023-11-10 00:19:30.000000]]
+               >
+               """
+               |> String.trim_trailing()
```
[the remark this answers is not in the document]
Comment (philss) [reply test/explorer/series/list_test.exs]: Nice! Thanks! 145c1e1f7fba7e7c03ed3fea81fb6bb22b56e74c

File: lib/explorer/shared.ex (hunk 3 above)
Comment (philss) [reply lib/explorer/shared.ex]: Good call! :D
I made the changes in 695a948147744b76a77745596ff7fdd32b179478

**Comment by billylanchantin** [comment]
Hey food for thought. This makes `DF.print` behave a little oddly.

```elixir
df_list = DF.new(a: [[1, 2], [3, 4]], b: [[5], [6, 7, 8, 9, 10, 11, 12, 13, 14, 15]])
DF.print(df_list)
# +-------------------------------------------+
# | Explorer DataFrame: [rows: 2, columns: 2] |
# +---------------------+---------------------+
# |          a          |          b          |
# |   <list[integer]>   |   <list[integer]>   |
# +=====================+=====================+
# | 1                   | 5                   |
# | 2                   |                     |
# +---------------------+---------------------+
# | 3                   | 6                   |
# | 4                   | 7                   |
# |                     | 8                   |
# |                     | 9                   |
# |                     | 10                  |
# |                     | 11                  |
# |                     | 12                  |
# |                     | 13                  |
# |                     | 14                  |
# |                     | 15                  |
# +---------------------+---------------------+
```

We may want to display it more like this:

```elixir
df_string = DF.new(a: ["[1, 2]", "[3, 4]"], b: ["[5]", "[6, 7, 8, 9, 10, 11, 12, 13, 14, 15]"])
DF.print(df_string)
# +-------------------------------------------------+
# |    Explorer DataFrame: [rows: 2, columns: 2]    |
# +----------+--------------------------------------+
# |    a     |                  b                   |
# | <string> |               <string>               |
# +==========+======================================+
# | [1, 2]   | [5]                                  |
# +----------+--------------------------------------+
# | [3, 4]   | [6, 7, 8, 9, 10, 11, 12, 13, 14, 15] |
# +----------+--------------------------------------+
```

I've actually been fiddling with making `DF.print` more configurable. While I don't think we should do anything about it on this PR, I'll be keeping it in mind as I play with possibilities.

**Comment by philss** [comment]
Thank you all! :purple_heart: 

@billylanchantin oh, good catch! I'm going to merge this one and we can fix that in another PR. Thanks!

**Pull request merged**